### PR TITLE
feat(markdown): exclude table column-length counting for threshold-exceeded rows

### LINKED ISSUE #4195: markdown files - let's limit excessive lines when prettying the markdown tables
Currently, there is no limit on the length of the contents of the cleaned markdown table columns. When we process tables which contain HTML, for example, all-contributors tables, the cleaned lines get very long:

long-lines

In the case above, line is 865 characters long.

Let's detect such cases and turn off the table normalising, falling back to usual three-dash pattern, `| :---: | :---: | :---: | :---: | :---: | :---: | :---: |`

Update.

Since we're using `all-contributors` spec tables as an example, we have to use a real-world markdown table for testing. I'm going to post the contributors table from one of my libraries here. Please don't consider this as a shameless plug, I first wanted to use all-contributors readme itself but later thought, technically, we should then mention its MIT copyright licence in Prettier's licence. Since I'm providing piece of my own library licence, licensing issues don't really apply.

Here is an original markdown example:

```md
<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
| [<img src="https://avatars1.githubusercontent.com/u/8344688?v=4" width="100px;"/><br /><sub><b>Roy Revelt</b></sub>](https://github.com/revelt)<br /> [💻](https://github.com/codsen/detergent/commits?author=revelt "Code") [📖](https://github.com/codsen/detergent/commits?author=revelt "Documentation") [⚠️](https://github.com/codsen/detergent/commits?author=revelt "Tests") | [<img src="https://avatars3.githubusercontent.com/u/1874682?v=4" width="100px;"/><br /><sub><b>Max Lapides</b></sub>](https://github.com/maxlapides)<br /> [🐛](https://github.com/codsen/detergent/issues?q=author%3Amaxlapides "Bug reports") | [<img src="https://avatars3.githubusercontent.com/u/9990521?v=4" width="100px;"/><br /><sub><b>Nacim Goura</b></sub>](https://github.com/nacimgoura)<br /> [🐛](https://github.com/codsen/detergent/issues?q=author%3Anacimgoura "Bug reports") | [<img src="https://avatars2.githubusercontent.com/u/23390212?v=4" width="100px;"/><br /><sub><b>Jack Steam</b></sub>](https://github.com/jacksteamdev)<br /> [🐛](https://github.com/codsen/detergent/issues?q=author%3Ajacksteamdev "Bug reports") |
| :---: | :---: | :---: | :---: |
<!-- ALL-CONTRIBUTORS-LIST:END -->
```

Here it is after being ran through Prettier:

```md
<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->

| [<img src="https://avatars1.githubusercontent.com/u/8344688?v=4" width="100px;"/><br /><sub><b>Roy Revelt</b></sub>](https://github.com/revelt)<br /> [💻](https://github.com/codsen/detergent/commits?author=revelt "Code") [📖](https://github.com/codsen/detergent/commits?author=revelt "Documentation") [⚠️](https://github.com/codsen/detergent/commits?author=revelt "Tests") | [<img src="https://avatars3.githubusercontent.com/u/1874682?v=4" width="100px;"/><br /><sub><b>Max Lapides</b></sub>](https://github.com/maxlapides)<br /> [🐛](https://github.com/codsen/detergent/issues?q=author%3Amaxlapides "Bug reports") | [<img src="https://avatars3.githubusercontent.com/u/9990521?v=4" width="100px;"/><br /><sub><b>Nacim Goura</b></sub>](https://github.com/nacimgoura)<br /> [🐛](https://github.com/codsen/detergent/issues?q=author%3Anacimgoura "Bug reports") | [<img src="https://avatars2.githubusercontent.com/u/23390212?v=4" width="100px;"/><br /><sub><b>Jack Steam</b></sub>](https://github.com/jacksteamdev)<br /> [🐛](https://github.com/codsen/detergent/issues?q=author%3Ajacksteamdev "Bug reports") |
| :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |


<!-- ALL-CONTRIBUTORS-LIST:END -->
```

Imho, ideally, Prettier should leave such table as it is, with three-dash separators (`:---:`) instead of long lines. Feel free to do it differently.

### PR DESCRIPTION
**Pull request by ikatyang**
Fixes #4195

~~HTML rows are usually very long long long, so excluding it should be a good solution I think.~~

~~Since the [all-contributors table](https://github.com/kentcdodds/all-contributors/blob/master/README.md#contributors) contains `html` in every row, I guess this solution should be fine,~~ the only difference is that our min separator is `:-:` instead of `:---:`.

Edit: Since [containing `html` is not a good feature](https://github.com/prettier/prettier/pull/4196#discussion_r176801859), I think excluding the row that'll cause the line exceeded threshold should be a better solution, but the threshold (current: `printWidth`) is arguable.

cc @revelt

### CODE REVIEW

File: tests/markdown/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -1693,8 +1693,8 @@ Built-in parsers:
 
 [Custom parsers](#custom-parser-api) are also supported. _Since v1.5.0_
 
-| Default   | CLI Override                                    | API Override                                               |
-| --------- | ----------------------------------------------- | ---------------------------------------------------------- |
+| Default | CLI Override | API Override |
+| ------- | ------------ | ------------ |
 | \`babylon\` | \`--parser <string>\`<br />\`--parser ./my-parser\` | \`parser: "<string>"\`<br />\`parser: require("./my-parser")\` |
```
Comment (j-f1) [inline tests/markdown/__snapshots__/jsfmt.spec.js.snap]: Is this change intentional?

Comment (ikatyang) [reply tests/markdown/__snapshots__/jsfmt.spec.js.snap]: Yeah, since there's a `html` and it's indeed very long.

Comment (j-f1) [reply tests/markdown/__snapshots__/jsfmt.spec.js.snap]: Looks like the first column got shorter even though it doesn’t have HTML in it.

Comment (revelt) [reply tests/markdown/__snapshots__/jsfmt.spec.js.snap]: That's expected, isn't it? There is no other column with would set the wider char count that would warrant that whitespace?!?

Comment (ikatyang) [reply tests/markdown/__snapshots__/jsfmt.spec.js.snap]: Since the table is row-based, we're not able to align only one column, though in this case it looks like we can but it's because it's the first column.

Comment (j-f1) [reply tests/markdown/__snapshots__/jsfmt.spec.js.snap]: Now I understand the logic here. LGTM 👍

File: tests/markdown_table/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -61,11 +61,31 @@ exports[`html.md 1`] = `
...
+| :-: | :-: | :-: | :-: |
+|  a  |  b  |  c  |  d  |
+
+| test                                    | test                          | test                          |
+| --------------------------------------- | ----------------------------- | ----------------------------- |
+| something something something something | something something           | something something something |
+| <a href="#">something something something something</a> | something something | something something something |
```
Comment (suchipi) [inline tests/markdown_table/__snapshots__/jsfmt.spec.js.snap]: In this case, the HTML is not particularly long, so the alignment looks odd. I think we might need a more fine-grained heuristic than just "contains HTML".

Comment (revelt) [reply tests/markdown_table/__snapshots__/jsfmt.spec.js.snap]: It's somewhat odd, but better than completely minimal column length, isn't it? I think one of the main points is comparing the length of the current row that contains HTML (edge-to-edge) in one cell on that row AND the length of the whole result, if we gave a green light and extended it.

ESLint's [max-len](https://eslint.org/docs/rules/max-len) default value is `80`. Therefore, it's a good guide - if any of the cleaned lines exceed `80` because of cleaning, that cleaning should be undo'ed.

What do you think?

Comment (suchipi) [reply tests/markdown_table/__snapshots__/jsfmt.spec.js.snap]: That sounds reasonable, but I think we should use printWidth instead of 80.

File: tests/markdown_table/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -61,11 +61,41 @@ exports[`html.md 1`] = `
...
+| xxx                     | xxx                    | xxx                      |
+| ----------------------- | ---------------------- | ------------------------ |
+| aaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbb | cccccccccccccccccccc     |
+| aaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbb | cccccccccccccccccccccccc |
+| aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbb | cccccccccccccccccccc |
+| aaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb | ccccc |
+| aaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbb | ccccccccccccccccccccccccccccccc |
+| aaaaaaaaaa              | bbbbbbbbbbbbbbbbbbbbbb | cccccccccccccccccccc     |
+| aaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbb            | cccccccccccccccccccc     |
+| aaaaaaaaaaaaaaaaaaaaaaa | bbbbbbbbbbbbbbbbbbbbbb | cccccccccc               |
```
Comment (suchipi) [inline tests/markdown_table/__snapshots__/jsfmt.spec.js.snap]: Hmm. This doesn't look very good.

**Comment by suchipi** [comment]
I think we should maybe not do this, and instead only add an exception for all-contributors tables (lipis said the same here https://github.com/prettier/prettier/issues/4195#issuecomment-375741268)

**Comment by ikatyang** [comment]
I'd like a more general solution, how about `<!-- prettier-ignore-start -->` + `<!-- prettier-ignore-end -->`?

**Comment by suchipi** [comment]
Is `prettier-ignore` not able to ignore an entire table?

**Comment by ikatyang** [comment]
It's able to ignore the entire table, but they have to put `<!-- prettier-ignore -->` in the auto-generated section, which means it'll be erased every time unless they auto-generate a `<!-- prettier-ignore -->`.

```md
<!-- some tool start -->

<!-- prettier-ignore -->
| some | table |
| ---- | ----  |
| some | table |
| some | table |

<!-- some tool end -->
```

```md
<!-- prettier-ignore-start -->
<!-- some tool start -->

| some | table |
| ---- | ----  |
| some | table |
| some | table |

<!-- some tool end -->
<!-- prettier-ignore-end -->
```

**Comment by ikatyang** [comment]
I guess we can only allow top-level `prettier-ignore-start/end` since this feature is aimed to ignore the auto-generated section, which should be always top-level, and also this can apply to other auto-generated section as well, e.g. TOC.

The `prettier-ignore-next` idea is good but I don't think people remember the name of every node, `start/end` is more intuitive.

**Comment by suchipi** [comment]
Restricting to top-level seems like a good solution :+1:

**Comment by ikatyang** [comment]
Opened #4202.
